In [1]:
%load_ext sql

Connect to the sql_project database on AWS RDS 

In [2]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

# Exploratory Queries

#1 Data quality check. 
Theoretically if the account is in the trial balance listing, it should have an ending balance for this year. The account would not appear unless action happened within this year

In [5]:
%%sql
SELECT 
    account_name, 
    t.account_number, 
    entry_number, 
    balance_forward, 
    ytd_debits, 
    ytd_credits, 
    ending_balance, 
    PY_ytd_balance 
FROM trial_balance t
JOIN account_names a
    ON t.account_number = a.account_number
WHERE ending_balance IS NULL; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[('Burden', '4540', 9454, 0, 7069, 4045, None, None)]

Found there is an account that has no ending balance either last year or this year but has debits and credits that happened this year. The debits and credits are also not even/balanced which is suspicious. This account could be fradulent or the data could be inaccurate. 

#1 continued for Zillow. Data quality checks 

In [5]:
%%sql 
SELECT 
    zpid, 
    address,
    price, 
    sqft 
FROM zillow 
WHERE price IS NULL; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
0 rows affected.


[]

No nulls found, the zillow data is complete. Makes sense since I pulled it from an API and not from the internet 

#2 check if the trial balance is balancing by summing the credits and debits. Also summing the this year ending balance and last year ending balance to see how they compare. 

In [6]:
%%sql
SELECT 
    SUM(ytd_debits) AS sum_debits, 
    SUM(ytd_credits) AS sum_credits, 
    SUM(ending_balance) AS sum_ending_balance,
    SUM(PY_ytd_balance) AS prior_year_sum_ending_balance
FROM trial_balance;

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[(Decimal('3427452698'), Decimal('3373152343'), Decimal('-25310524'), Decimal('-23292879'))]

The debits and credits do not balance so there must be some errors within the data or trial balance and auditors would need to investigate that. Comparing last year to this year is another common procedure auditors do to make sure this year's number does not look abnormally large (depending on what has changed in the business environment too). The ending balance comparison looks reasonable since they are both around 23-25 million. 

#3 Comparing the relationship and checking Accounts receivable and Allowance for uncollectible accounts is important since these two accounts are directly related. Allowance for uncollectible accounts should be reasonably proportionate to accounts receivable. 

In [15]:
%%sql
SELECT
    t.account_number, 
    account_name, 
    balance_forward, 
    ytd_debits, 
    ytd_credits,
    ending_balance,
    PY_ytd_balance
FROM trial_balance t
JOIN account_names a
    ON t.account_number = a.account_number
WHERE t.account_number IN (2218, 2219)
GROUP BY t.account_number;

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
2 rows affected.


[('2218', 'Accounts Receivable', 873964, 6093882, 4337829, 2630017, 1446972),
 ('2219', 'Allowance For Uncoll Accts', -397234, 0, 0, -397234, -152642)]

This verifies the relationship between A/R and Allowance for uncollectible accounts is reasonable. Allowance for doubtful accounts roughly doubled but A/R also increased. We could further investigate why it seems they wrote off more uncollectibles this year than last if we felt we needed to. 

#4 Using Min and Max to examine certain accounts and entries. Using this on accounts payable to see small/large entries for owing people money at the end of the year. 

In [30]:
%%sql
SELECT
    t.account_number, 
    account_name, 
    PY_ytd_balance, 
    MIN(ending_balance),
    MAX(ending_balance)
FROM trial_balance t
JOIN account_names a
    ON t.account_number = a.account_number
WHERE t.account_number = 4001;

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[('4001', 'Accounts Payable', 0, -401604, 180)]

This shows 401604 is a very large number to have as an ending balance in the liability account so we might want to examine this specific entry and what department it goes to. 180 on the other hand is also odd for a liabilty account to be positive but it could be an entry paying off liability so could look into but not much worry. 

#4 for zillow. Finding the min and max could be interesting to get a range and make sure there are not any outliers to investigate. 

In [17]:
%%sql 
SELECT 
    MIN(price) AS MinPrice, 
    MAX(price) AS MaxPrice 
FROM zillow; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[('10000000', '9999999')]

The max and min prices are very similar which shows its a tight market but also that there are no outliers or weird data in the set. You could also do this on specific zipcodes if you wanted. 

#5 Looking at depreciation relationships. There should be a relationship between depreciation and a contra asset account like accumulated depreciation. This finds our depreciation accounts for us. 

In [14]:
%%sql
SELECT 
    t.account_number, 
    account_name, 
    balance_forward, 
    ytd_debits, 
    ytd_credits,
    ending_balance,
    PY_ytd_balance
FROM trial_balance t 
JOIN account_names a
    ON t.account_number = a.account_number
WHERE account_name LIKE '%depreciation%'; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
96 rows affected.


[('2423', 'Fleet Depreciation', 0, 0, 0, 0, 3392),
 ('2423', 'Fleet Depreciation', 0, 0, 0, 0, 5168),
 ('2423', 'Fleet Depreciation', 0, 25392, 0, 25392, 61080),
 ('2423', 'Fleet Depreciation', 0, 2768, 0, 2768, 3688),
 ('2423', 'Fleet Depreciation', 0, 146704, 0, 146704, 31000),
 ('2423', 'Fleet Depreciation', 0, 8128, 0, 8128, 9752),
 ('2423', 'Fleet Depreciation', 0, 8864, 0, 8864, 0),
 ('2423', 'Fleet Depreciation', 0, 1088, 0, 1088, 0),
 ('2423', 'Fleet Depreciation', 0, 26376, 0, 26376, 19552),
 ('2423', 'Fleet Depreciation', 0, 19128, 0, 19128, 11496),
 ('2423', 'Fleet Depreciation', 0, 211696, 0, 211696, 83624),
 ('2423', 'Fleet Depreciation', 0, 3880, 0, 3880, 3232),
 ('2423', 'Fleet Depreciation', 0, 3088, 0, 3088, 3088),
 ('2423', 'Fleet Depreciation', 0, 11392, 0, 11392, 11392),
 ('2423', 'Fleet Depreciation', 0, 102040, 0, 102040, 105064),
 ('2423', 'Fleet Depreciation', 0, 52016, 0, 52016, 19960),
 ('2423', 'Fleet Depreciation', 0, 1296, 0, 1296, 768),
 ('2423', 'Fleet Depreciation', 0, 7480, 0, 7480, 7792),
 ('2423', 'Fleet Depreciation', 0, 9776, 0, 9776, 9776),
 ('2423', 'Fleet Depreciation', 0, 12024, 0, 12024, 7720),
 ('2423', 'Fleet Depreciation', 0, 19936, 0, 19936, 19936),
 ('2423', 'Fleet Depreciation', 0, 56856, 0, 56856, 47752),
 ('2423', 'Fleet Depreciation', 0, 18016, 0, 18016, 9928),
 ('2423', 'Fleet Depreciation', 0, 344, 0, 344, 344),
 ('4100', 'Depreciation', 0, 1314, 0, 1314, 1502),
 ('2423', 'Fleet Depreciation', 0, 10488, 0, 10488, 5760),
 ('4100', 'Depreciation', 0, 1351424, 0, 1351424, 1611859),
 ('4100', 'Depreciation', 0, 105044, 0, 105044, 113144),
 ('2423', 'Fleet Depreciation', 0, 73056, 0, 73056, 71777),
 ('2423', 'Fleet Depreciation', 0, 148208, 0, 148208, 4896),
 ('2423', 'Fleet Depreciation', 0, 15256, 0, 15256, 15256),
 ('2423', 'Fleet Depreciation', 0, 2400, 0, 2400, 2400),
 ('2423', 'Fleet Depreciation', 0, 2184, 0, 2184, 2184),
 ('2423', 'Fleet Depreciation', 0, 4832, 0, 4832, 4832),
 ('2423', 'Fleet Depreciation', 0, 328, 0, 328, 328),
 ('2423', 'Fleet Depreciation', 0, 1944, 0, 1944, 1944),
 ('2423', 'Fleet Depreciation', 0, 1968, 0, 1968, 1968),
 ('2423', 'Fleet Depreciation', 0, 14672, 0, 14672, 7207),
 ('2423', 'Fleet Depreciation', 0, 212392, 0, 212392, 204384),
 ('2423', 'Fleet Depreciation', 0, 30936, 0, 30936, 30936),
 ('2423', 'Fleet Depreciation', 0, 8776, 0, 8776, 8776),
 ('2423', 'Fleet Depreciation', 0, 18768, 0, 18768, 17216),
 ('2423', 'Fleet Depreciation', 0, 120616, 0, 120616, 120616),
 ('2423', 'Fleet Depreciation', 0, 9632, 0, 9632, 5600),
 ('2423', 'Fleet Depreciation', 0, 18816, 0, 18816, 17808),
 ('2423', 'Fleet Depreciation', 0, 53096, 0, 53096, 20384),
 ('2423', 'Fleet Depreciation', 0, 544, 0, 544, 544),
 ('2423', 'Fleet Depreciation', 0, 3720, 0, 3720, 3720),
 ('2423', 'Fleet Depreciation', 0, 61584, 0, 61584, 68416),
 ('2423', 'Fleet Depreciation', 0, 69688, 0, 69688, 64536),
 ('2423', 'Fleet Depreciation', 0, 3928, 0, 3928, 3928),
 ('2423', 'Fleet Depreciation', 0, 59048, 0, 59048, 29528),
 ('4100', 'Depreciation', 0, 305146, 0, 305146, 347879),
 ('2423', 'Fleet Depreciation', 0, 24976, 0, 24976, 24288),
 ('4100', 'Depreciation', 0, 2955617, 0, 2955617, 2877756),
 ('2423', 'Fleet Depreciation', 0, 2752, 0, 2752, 21912),
 ('2423', 'Fleet Depreciation', 0, 40256, 0, 40256, 29824),
 ('2423', 'Fleet Depreciation', 0, 21744, 0, 21744, 15136),
 ('2423', 'Fleet Depreciation', 0, 10880, 0, 10880, 9864),
 ('2423', 'Fleet Depreciation', 0, 68312, 0, 68312, 19608),
 ('2423', 'Fleet Depreciation', 0, 0, 0, 0, 1016),
 ('2423', 'Fleet Depreciation', 0, 28696, 0, 28696, 9656),
 ('2423', 'Fleet Depreciation', 0, 3432, 0, 3432, 2032),
 ('4100', 'Depreciation', 0, 4495773, 0, 4495773, 4948914),
 ('4100', 'Depreciation', 0, 984570, 0, 984570, 1125224),
 ('4100', 'Depreciation', 0, 41337, 0, 41337, 47540),
 ('2423', 'Fleet Depreciation', 0, 5968, 0, 5968, 2144),
 ('2423', 'Fleet Depreciation', 0, 9416, 0, 9416, 9416),
 ('4100', 'Depreciation', 0, 13404, 0, 13404, 13036)

This is concerning there is no credit account account associated with the debit depreciation account. There should be a credit to accumulated depreciation so it balances. We would need to look into where these credits are going like if they are going directly off the equipment rather than a contra asset account or if the data is incomplete. 

# Business Question Queries 

Directions: To answer your 3 questions of the data (1 primary and 2 related to the primary), you must use all 7 of the SQL features below. Features can be combined into a single query.

1.	GROUP BY 
2.	JOIN 
3.	CASE
4.	VIEW 
5.	Common table expression (CTE) 
6.	Subquery
7.	Window function 

In a Markdown cell above each SQL statement, provide the question, the business justification, and the SQL feature(s) used. In a Markdown cell below the SQL query's results, give the recommendation based on the results.

VIEW Syntax
CREATE [OR REPLACE] VIEW view_name AS
    SELECT columns
    FROM tables
    [WHERE conditions];
    
Window function can use a rank or LAG 

# 1) Primary business question: From an auditing perspective, are the valuations of the Trial Balance accounts (or for zillow the houses) what we expect?

In accounting we have multiple valuation methods depending on what guidelines you follow and other factors, historical cost or market valuation. Historical cost is valuing a buidling/asset at the cost it was purchased for (often years ago) while market valuation is what you could sell the asset for right now. If we had a house or building on our financial statements, the auditor could inspect zillow data to see the market price for similar buildings in the area. 

For this example, assuming we have a building in the 90077 zipcode around 15,000 sqft. 

This uses a subquery to extract addresses that are under these qualifications. It also uses a group by to be able to make sure it only shows one entry per address. 

In [18]:
%%sql 
SELECT 
    address, 
    sqft, 
    price
FROM zillow 
WHERE address IN 
    (SELECT address 
     FROM zillow 
     WHERE sqft BETWEEN 13000 AND 17000
        AND address LIKE '%90077')
GROUP BY address
ORDER BY price; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
4 rows affected.


[('10480 W Sunset Blvd, Los Angeles, CA 90077', '13000', '19995000'),
 ('948 Bel Air Rd, Los Angeles, CA 90077', '17000', '19995000'),
 ('670 Perugia Way, Los Angeles, CA 90077', '14282', '38500000'),
 ('111 N Mapleton Dr, Los Angeles, CA 90077', '14259', '43750000')]

This gives the market prices of comparable buildings in the 90077 area. From this you can see a building can be the same price even at 4,000 gap in square footage. The recommendation from here would be to compare your building on the financial statement to the price of these buildings to see if it is comparable. Also might have to look into other real estate factors for the valuation of the building but this is a great starting point. 

# 2) Are the zillow listing comparable to each other? To use in valuation analysis

For the houses around 15,000 square footage in the 90077 area that we found in the above query, we could look at the max, min, and average price of them compared to each property to see what a better estimate of value would be. Some houses might be over or undervalued and not a good basis of our valuation. 

This uses a window function to be able to display the average home price, min and max home price, next to each home entry and price. 

In [15]:
%%sql
SELECT 
    address,
    price, 
    AVG(price) OVER () AS avg_price_for_houses, 
    price < AVG(price) OVER () AS price_less_than_avg_price_for_houses,
    MIN(price) OVER () AS min_price_for_houses,
    MAX(price) OVER () AS max_price_for_houses
FROM zillow
WHERE address LIKE '%90077'
    AND sqft BETWEEN 13000 AND 17000
GROUP BY address; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
4 rows affected.


[('10480 W Sunset Blvd, Los Angeles, CA 90077', '19995000', 30560000.0, 1, '19995000', '43750000'),
 ('111 N Mapleton Dr, Los Angeles, CA 90077', '43750000', 30560000.0, 0, '19995000', '43750000'),
 ('670 Perugia Way, Los Angeles, CA 90077', '38500000', 30560000.0, 0, '19995000', '43750000'),
 ('948 Bel Air Rd, Los Angeles, CA 90077', '19995000', 30560000.0, 1, '19995000', '43750000')]

This query shows there is a range of prices in the houses in the zip code and two of the houses are a lot more expensive than the average price. I would recommend we take the two houses priced at $19995000 out of the data we use for valuation because it will skew it. 

When looking at comparability, we need to make sure the zillow listings are active. If the house has been on the market for too long and just sitting on zillow then it might not be a good data point to use in valuation. In this case, we are going to look at listings that have been on there for a day. 

This uses the VIEW function to be able to create this view for active listings. 

In [3]:
%%sql
CREATE OR REPLACE VIEW active_zillow_listing AS
    SELECT * 
    FROM zillow 
    WHERE days_on_zillow = -1;

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
0 rows affected.


[]

In [4]:
%%sql
SELECT * 
FROM active_zillow_listing; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1095 rows affected.


[(1, '594 S Mapleton Dr, Los Angeles, CA 90024', '165000000', '56500', '4.6017', '-1'),
 (2, '4050 Contera Rd, Encino, CA 91436', '6479000', '6297', '0.33434343434343', '-1'),
 (3, '9570 Hidden Valley Rd, Beverly Hills, CA 90210', '19475000', '5427', '1.1639', '-1'),
 (4, '1332 Laurel Way, Beverly Hills, CA 90210', '28950000', '11220', '0.4001', '-1'),
 (5, '10979 Chalon Rd, Los Angeles, CA 90077', '68500000', '32000000', '1.0293', '-1'),
 (6, '4565 Encino Ave, Encino, CA 91316', '12995000', '12300', '0.50955004591368', '-1'),
 (7, '9332 Hazen Dr, Beverly Hills, CA 90210', '5995000', '5183', '9909.9', '-1'),
 (8, '441 N Bristol Ave, Los Angeles, CA 90049', '23500000', '9385', '0.4496', '-1'),
 (9, '10066 Cielo Dr, Beverly Hills, CA 90210', '85000000', '21000', '3.6007', '-1'),
 (10, '4430 Hayvenhurst Ave, Encino, CA 91436', '9480000', '7829', '0.7187', '-1'),
 (11, '2235 Hercules Dr, Los Angeles, CA 90046', '7995000', '6162', '0.3163', '-1'),
 (12, '4838 Calderon Rd, Woodland Hills, CA 91364', '12999000', '13379', '0.6344', '-1'),
 (13, '2571 Wallingford Dr, Beverly Hills, CA 90210', '85000000', '38000', '5.2', '-1'),
 (14, '1369 Londonderry Pl, Los Angeles, CA 90069', '37995000', '14000', '0.4288', '-1'),
 (15, '101 S Larchmont Blvd, Los Angeles, CA 90004', '4195000', '3180', '6926.04', '-1'),
 (16, '1714 Stone Canyon Rd, Los Angeles, CA 90077', '20800000', '17415', '4.7716', '-1'),
 (17, '1120 Angelo Dr, Beverly Hills, CA 90210', '5500000', '5138', '0.322', '-1'),
 (18, '1420 Davies Dr, Beverly Hills, CA 90210', '87000000', '21800', '7.8', '-1'),
 (19, '10250 W Sunset Blvd, Los Angeles, CA 90077', '75000000', '30000', '2.8137', '-1'),
 (20, '643 Radcliffe Ave, Pacific Palisades, CA 90272', '5395000', '3677', '0.2706', '-1'),
 (21, '5531 Red Oak Dr, Los Angeles, CA 90068', '4599000', '3960', '7278.876', '-1'),
 (22, '10771 Bellagio Rd, Los Angeles, CA 90077', '65000000', '19721', '1.1949', '-1'),
 (23, '16366 Sloan Dr, Los Angeles, CA 90049', '5199000', '4600', '0.3489', '-1'),
 (24, '1945 Linda Flora Dr, Los Angeles, CA 90077', '10999988', '7640', '0.3973', '-1'),
 (25, '315 S Hudson Ave, Los Angeles, CA 90020', '21950000', '13240', '0.696', '-1'),
 (26, '846 S Longwood Ave, Los Angeles, CA 90005', '5999000', '4508', '0.335', '-1'),
 (27, '1640 Stradella Rd, Los Angeles, CA 90077', '14995000', '9246', '0.9821', '-1'),
 (28, '15321 Kingswood Ln, Sherman Oaks, CA 91403', '8995000', '9075', '0.47288797061524', '-1'),
 (29, '9650 Cedarbrook Dr, Beverly Hills, CA 90210', '250000000', '78000', '36.7309', '-1'),
 (30, '271 S Mapleton Dr, Los Angeles, CA 90024', '64950000', '31450', '1.4021', '-1'),
 (31, '15000 Corona Del Mar, Pacific Palisades, CA 90272', '29995000', '6673', '0.7758', '-1'),
 (32, '19200 Rosita St, Tarzana, CA 91356', '5880000', '5289', '0.4549', '-1'),
 (33, '4048 Sunnyslope Ave, Sherman Oaks, CA 91423', '5499000', '5405', '0.2935', '-1'),
 (34, '4453 Balboa Ave, Encino, CA 91316', '6795000', '7500', '1.0116', '-1'),
 (35, '746 S Westholme Ave, Los Angeles, CA 90024', '6750000', '4504', '9356.688', '-1'),
 (36, '4651 Louise Ave, Encino, CA 91316', '6298500', '6114', '0.42371441689624', '-1'),
 (37, '9814 Curwood Pl, Beverly Hills, CA 90210', '12868000', '9500', '0.7725', '-1'),
 (38, '9726 Hensal Rd, Beverly Hills, CA 90210', '6495000', '4768', '0.292', '-1'),
 (39, '16041 Woodvale Rd, Encino, CA 91436', '26000000', '19500', '1.265', '-1'),
 (40, '19333 Rosita St, Tarzana, CA 91356', '7495000', '9738', '0.67063820018365', '-1'),
 (41, '594 S Mapleton Dr, Los Angeles, CA 90024', '165000000', '56500', '4.6017', '-1'),
 (42, '4050 Contera Rd, Encino, CA 91436', '6479000', '6297', '0.33434343434343', '-1'),
 (43, '9570 Hidden Valley Rd, Beverly Hills, CA 90210', '19475000', '5427', '1.1639', '-1'),
 (44, '1332 Laurel Way, Beverly Hills, CA 90210', '28950000', '11220', '0.4001', '-1'),
 (45, '10979 Chalon Rd, Los Angeles, CA 90077', '68500000', '32000000', '1.0293', '-1'),
 (46, '4565 Encino Ave, Encino, CA 91316', '1

This query shows we have 1095 entries that are active on the market. I recommend we only use these entries in our valuation calculations

# Primary business question of valuation applying to the trial balance account: accounts receivable and allowance for uncollectible accounts reasonable?

An account we can check to see if it is valued at what we would expect is the allowance for uncollectible accounts. Usually the amount of accounts receivable that becomes uncollectible is consistent year over year or is based on a percentage. For example, allowance for uncollectible accounts could be 10% of accounts receivable. 

This uses A CTE to be able to sum the A/R account and the uncollectible account for prior and current year and then calculate the percent change in uncollectible rate. Within the CTEs we also use a join to be able to show the account name along with the account number and a group by to sum up by account number. 

In [8]:
%%sql
WITH AR_total AS (
    SELECT
        t.account_number, 
        account_name, 
        balance_forward, 
        ytd_debits, 
        ytd_credits,
        ending_balance AS ending_AR,
        PY_ytd_balance AS prior_year_AR
    FROM trial_balance t
    JOIN account_names a
        ON t.account_number = a.account_number
    WHERE t.account_number = 2218
    GROUP BY t.account_number
), 
Uncollectible_total AS (
    SELECT
        t.account_number, 
        account_name, 
        balance_forward, 
        ytd_debits, 
        ytd_credits,
        ending_balance AS ending_uncollectible,
        PY_ytd_balance AS prior_year_uncollectible
    FROM trial_balance t
    JOIN account_names a
        ON t.account_number = a.account_number
    WHERE t.account_number = 2219
    GROUP BY t.account_number
)
SELECT  
    (ending_uncollectible / ending_AR) AS current_year_uncollectible_rate, 
    (prior_year_uncollectible / prior_year_AR) AS prior_year_uncollectible_rate
FROM AR_total, Uncollectible_total; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[(Decimal('-0.1510'), Decimal('-0.1055'))]

Based on these results the auditor would want to investigate why their uncollectible rate went up a full five percentage points and overall increased by 50%. Even if the overall A/R increased, it is suspicious the overall uncollectible rate increased by 5% from year to year. They could further look into if sales increased, if a major customer went out of business and could not pay them back, and overall what customers they sold product/services to that caused the increase in uncollectible amount. 

Often in auditing we compare the prior year balance to the current year balance to determine if the current year balance is reasonable. With using a case when, we can easily flag which account's balance are above/below the prior year balances by having a column saying so next to each account. It also uses a join so we can display the account name with the account number.

In [9]:
%%sql
SELECT 
    t.account_number, 
    account_name,  
    ending_balance, 
    PY_ytd_balance, 
    (ending_balance - PY_ytd_balance) AS difference, 
    CASE 
        WHEN ending_balance > PY_ytd_balance THEN 'Current Year Greater than Prior Year'
        WHEN ending_balance < PY_ytd_balance THEN 'Current Year Less than Prior Year'
        ELSE 'Equal or N/A'
    END AS Current_Versus_Prior_Balance 
FROM trial_balance t 
JOIN account_names a
    ON t.account_number = a.account_number
GROUP BY t.account_number
ORDER BY difference; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
1168 rows affected.


[('5551.0059', 'Deferred Revenue Deferred Inflow of Pension', -62719658, -722862, -61996796, 'Current Year Less than Prior Year'),
 ('1129', 'Taxes Collected For Other Governments', -284781289, -225842198, -58939091, 'Current Year Less than Prior Year'),
 ('1289', 'NTAM EAFE', 20061975, 48079212, -28017237, 'Current Year Less than Prior Year'),
 ('2256', 'Amt For Debt Svc Gen Oper', 200384879, 226950803, -26565924, 'Current Year Less than Prior Year'),
 ('5551.0052', 'Deferred Revenue Deferred Inflow of OPEB', -28800016, -13591573, -15208443, 'Current Year Less than Prior Year'),
 ('2400.0010', 'Equity In 0010', -65883114, -52031307, -13851807, 'Current Year Less than Prior Year'),
 ('5553', 'Unearned Revenue', -12091315, 0, -12091315, 'Current Year Less than Prior Year'),
 ('1297', 'Hexavest', 0, 10254983, -10254983, 'Current Year Less than Prior Year'),
 ('2400.0042', 'Equity In 0042', -80823603, -72172389, -8651214, 'Current Year Less than Prior Year'),
 ('1006', 'Cash Treasurer Current Tax', 2046684, 9726742, -7680058, 'Current Year Less than Prior Year'),
 ('2400.0043', 'Equity In 0043', -64298329, -57587756, -6710573, 'Current Year Less than Prior Year'),
 ('1346', 'Affordable Housing Levy', -6326196, 0, -6326196, 'Current Year Less than Prior Year'),
 ('1217', 'Cash Ret (Northern Trust)', 2630945, 8377689, -5746744, 'Current Year Less than Prior Year'),
 ('2400.0103', 'Equity In 0103', -5718313, 0, -5718313, 'Current Year Less than Prior Year'),
 ('6682', 'Contributed Assets Capital Projects', -70627030, -65411242, -5215788, 'Current Year Less than Prior Year'),
 ('2266.0059', 'Deferred Outflow of Resources Deferred Outflow', 5192423, 10342749, -5150326, 'Current Year Less than Prior Year'),
 ('2160', 'Federal Emerg Mgt Agcy - Grant', -2160440, 2073565, -4234005, 'Current Year Less than Prior Year'),
 ('1290', 'Constitution Capt’l', 152845, 3929496, -3776651, 'Current Year Less than Prior Year'),
 ('2400.0072', 'Equity In 0072', -32754542, -29444069, -3310473, 'Current Year Less than Prior Year'),
 ('2400.00CP', 'Equity In 00CP', -5274302, -2053539, -3220763, 'Current Year Less than Prior Year'),
 ('2400.0062', 'Equity In 0062', -18962630, -15746634, -3215996, 'Current Year Less than Prior Year'),
 ('1298', 'WCM', 6500000, 9500000, -3000000, 'Current Year Less than Prior Year'),
 ('2400.0012', 'Equity In 0012', -18252657, -15621966, -2630691, 'Current Year Less than Prior Year'),
 ('2400.0069', 'Equity In 0069', -22757952, -20173426, -2584526, 'Current Year Less than Prior Year'),
 ('2430', 'Contracted Services', 27110, 2216804, -2189694, 'Current Year Less than Prior Year'),
 ('7111', 'Metered Service', -19194529, -17215277, -1979252, 'Current Year Less than Prior Year'),
 ('2292', 'Prepaid Insurance', 254597, 2106884, -1852287, 'Current Year Less than Prior Year'),
 ('7145', 'Capital Recovery Charge', -2475648, -713185, -1762463, 'Current Year Less than Prior Year'),
 ('2400.0021', 'Equity In 0021', -17555293, -16041087, -1514206, 'Current Year Less than Prior Year'),
 ('1024', 'Michigan CLASS DDA Bond 2019-A', 4353478, 5856726, -1503248, 'Current Year Less than Prior Year'),
 ('2266.0052', 'Deferred Outflow of Resources Deferred Outflow', 16737554, 18206768, -1469214, 'Current Year Less than Prior Year'),
 ('2R00.2020', 'Taxes Rec - Real Property 2020', 0, 1356021, -1356021, 'Current Year Less than Prior Year'),
 ('1345', 'Sidewalk Construction Levy', -1265123, 0, -1265123, 'Current Year Less than Prior Year'),
 ('2272', 'State Shared - Constitutional', -8220969, -7009211, -1211758, 'Current Year Less than Prior Year'),
 ('4110', 'Principal', 1009987, 2139000, -1129013, 'Current Year Less than Prior Year'),
 ('2800.0010', 'Fringe Transfer 0010', -3957174, -2830566, -1126608, 'Current Year Less than Prior Year'),
 ('2257', 'Amt For Debt Svc Sp Rev Fund', 11513460, 12626460, -1113000, 'Current Year Less than Prior Year'),
 ('2400.0102', 'Equity In 0102', -1067127, 2762, -1069889, 'Current Year Less than Prior Year'),

The query orders by which accounts have the greatest difference so I would look into those with a material difference that we find alarming from this query. For revenues we would find it more concerning if it was greater than prior year while liabilities if they are less than prior year. We can use this query to detect if we think accounts are not accurate or complete. 

# 3) Is the company's payroll accurate? Is the contracted payroll accurate? 

Since payroll is often the biggest expense for companies, it is key to look at when auditing to make sure it is complete and accurate. They often confirm timesheets and other documents to confirm the accuracy of payroll expense. 
We can look at the relationships between the different types of payroll the company offers: permanent vs temporary vs contracted services. From an auditing perspective we could look into why things are contracted and if these contracts are accurate. 

This just uses a join and group by to specify which account numbers we want the totals for. 

In [15]:
%%sql
SELECT 
    t.account_number, 
    account_name, 
    balance_forward, 
    ytd_debits, 
    ytd_credits,
    ending_balance,
    PY_ytd_balance
FROM trial_balance t 
JOIN account_names a
    ON t.account_number = a.account_number
WHERE t.account_number IN (1100, 1200, 2430, 2431) 
GROUP BY t.account_number
ORDER BY ending_balance DESC; 

 * mysql://admin:***@lmu-dev-01.cgyzxqt5pvu5.us-east-1.rds.amazonaws.com/sql_project
4 rows affected.


[('2430', 'Contracted Services', 0, 44110, 17000, 27110, 2216804),
 ('1100', 'Permanent Time Worked', 0, 8582, 597, 7984, 11682),
 ('2431', 'Contracted Temp Empl Services', 0, 2103, 0, 2103, 0),
 ('1200', 'Temporary Pay', 0, 0, 0, 0, 13)]

Found that the entity contracts majority of their services which is interesting and something to look into why they have so much more contracted services. We should confirm the amounts with the actual contracts and confirm the normal payroll with timesheets.